lib/statsd/instrument/threaded_udp_sink.rb

```diff
@@ -0,0 +1,110 @@
+# frozen_string_literal: true
+
+module StatsD
+  module Instrument
+    # @note This class is part of the new Client implementation that is intended
+    #   to become the new default in the next major release of this library.
+    class ThreadedUDPSink
+      def self.for_addr(addr)
+        host, port_as_string = addr.split(":", 2)
+        new(host, Integer(port_as_string))
+      end
+
+      class Dispatcher
+        class << self
+          def finalizer(instance)
+            proc { instance.stop }
+          end
+        end
+
+        def initialize(host, port, queue)
+          @host = host
+          @port = port
+          @queue = queue
+          @socket = nil
+          @thread = nil
+          ObjectSpace.define_finalizer(self, self.class.finalizer(self))
```

Comment:
Oh, and there was also an issue with the finalizer, it would never fire because of that common gotcha: https://www.mikeperham.com/2010/02/24/the-trouble-with-ruby-finalizers/